src/asyncresults.jl

```diff
@@ -0,0 +1,256 @@
+        catch err
+            push!(errors, err)
+        end
+    end
+
+    if throw_error && !isempty(errors)
+        throw(CompositeException(errors))
+    elseif result === nothing
+        error(LOGGER, "Async query did not return result")
+    else
+        return result
+    end
+end
+
+function _consume(jl_conn::Connection)
+    async_result = jl_conn.async_result
+    result_ptrs = Ptr{libpq_c.PGresult}[]
+    watcher = FDWatcher(socket(jl_conn), true, false)  # can wait for reads
+    try
+        while true
+            if async_result.should_cancel
+                debug(LOGGER, "Received cancel signal for connection $(jl_conn.conn)")
+                _cancel(jl_conn)
+            end
+            debug(LOGGER, "Waiting to read from connection $(jl_conn.conn)")
+            wait(watcher)
+            debug(LOGGER, "Consuming input from connection $(jl_conn.conn)")
+            success = libpq_c.PQconsumeInput(jl_conn.conn) == 1
+            !success && error(error_message(jl_conn))
+
+            while libpq_c.PQisBusy(jl_conn.conn) == 0
+                debug(LOGGER, "Checking the result from connection $(jl_conn.conn)")
+                result_ptr = libpq_c.PQgetResult(jl_conn.conn)
+                if result_ptr == C_NULL
+                    debug(LOGGER, "Finished reading from connection $(jl_conn.conn)")
+                    return result_ptrs
+                else
+                    result_num = length(result_ptrs) + 1
+                    debug(LOGGER,
+                        "Saving result $result_num from connection $(jl_conn.conn)"
+                    )
+                    push!(result_ptrs, result_ptr)
+                end
+            end
+        end
+    finally
+        close(watcher)
+    end
+end
+
+"""
+    cancel(async_result::AsyncResult)
+
+If this [`AsyncResult`](@ref) represents a currently-executing query, attempt to cancel it.
+"""
+function cancel(async_result::AsyncResult)
+    async_result.should_cancel = true
+    return
+end
+
+function _cancel(jl_conn::Connection)
+    cancel_ptr = libpq_c.PQgetCancel(jl_conn.conn)
+    try
+        # https://www.postgresql.org/docs/10/libpq-cancel.html#LIBPQ-PQCANCEL
+        errbuf_size = 256
+        errbuf = fill(0x0, errbuf_size)
+        success = libpq_c.PQcancel(cancel_ptr, pointer(errbuf), errbuf_size) == 1
+        if !success
+            warn(LOGGER, "Failed cancelling query: $(String(errbuf))")
+        else
+            debug(LOGGER, "Cancelled query for connection $(jl_conn.conn)")
+        end
+    finally
+        libpq_c.PQfreeCancel(cancel_ptr)
+    end
+end
+
+iserror(async_result::AsyncResult) = Base.istaskfailed(async_result.result_task)
+Base.isready(async_result::AsyncResult) = istaskdone(async_result.result_task)
+Base.wait(async_result::AsyncResult) = wait(async_result.result_task)
+Base.fetch(async_result::AsyncResult) = fetch(async_result.result_task)
+trywait(async_result::AsyncResult) = (try wait(async_result) catch end; nothing)
```

Comment:
- What is this for:
 - Do we need to catch **everything** ?
 - should it be returrning `true` if the wait was done and `false` if an error was thrown?

Reply:
I think this can be removed now that it's no longer used internally. I'll make it a function in the test file.

In any case `wait` shouldn't return and neither should this.